pkg/epp/datalayer/metrics/datasource.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/url"
+	"strconv"
+	"sync"
+	"sync/atomic"
+
+	"sigs.k8s.io/gateway-api-inference-extension/pkg/epp/datalayer"
+)
+
+const (
+	datasourceName = "metrics-data-source"
+)
+
+// DataSource is the metrics data source, returning Prometheus formatted
+// metrics from an endpoint.
+type DataSource struct {
+	mspScheme string
+	mspPort   atomic.Pointer[string]
+	mspPath   string
+
+	clients    ClientFactory
+	extractors sync.Map // key: name, value: extractor
+}
+
+// NewDataSource returns a new metrics data source, configured with the provided
+// client factory. If ClientFactory is nil, a default factory is used.
+func NewDataSource(metricsScheme string, metricsPort int32, metricsPath string, clf ClientFactory) *DataSource {
+	if clf == nil {
+		clf = GetDefaultClientFactory()
+	}
+	ds := &DataSource{
```

Comment:
nit: specific in this case we might prefer using explicit names (e.g., dataSource or dataSrc).
ds seems a bit overloaded term (used when using Datastore)